src/probnum/problems/zoo/linalg/_suitesparse_matrix.py

```diff
@@ -0,0 +1,330 @@
+    psym :
+        Degree of symmetry of the matrix pattern.
+    nsym :
+        Degree of numerical symmetry of the matrix.
+    kind  :
+        Information of the problem domain this matrix arises from.
+
+    References
+    ----------
+    .. [1] Davis, TA and Hu, Y. The University of Florida sparse matrix
+           collection. *ACM Transactions on Mathematical Software (TOMS)* 38.1 (
+           2011): 1-25.
+    .. [2] Kolodziej, Scott P., et al. The SuiteSparse matrix collection website
+           interface. *Journal of Open Source Software* 4.35 (2019): 1244.
+    """
+
+    # pylint: disable="too-many-instance-attributes,too-many-arguments,abstract-method"
+    def __init__(
+        self,
+        matid: str,
+        group: str,
+        name: str,
+        dtype: DTypeArgType,
+        nrows: int,
+        ncols: int,
+        nnz: int,
+        is2d3d: bool,
+        isspd: bool,
+        psym: float,
+        nsym: float,
+        kind: str,
+    ):
+        self.matrix = None
+        self.matid = matid
+        self.group = group
+        self.name = name
+        self.nnz = nnz
+        self.is2d3d = is2d3d
+        self.isspd = isspd
+        self.psym = psym
+        self.nsym = nsym
+        self.kind = kind
+
+        super().__init__(shape=(nrows, ncols), dtype=np.dtype(dtype))
+
+    @classmethod
+    def from_database_entry(cls, database_entry: Dict) -> "SuiteSparseMatrix":
+        """Create a SuiteSparseMatrix object from an entry of the database index.
+
+        Parameters
+        ----------
+        database_entry :
+            Dictionary representing one entry from the SuiteSparse database index.
+        """
+        if bool(int(database_entry["logical"])):
+            dtype = np.dtype(np.bool_)
+        elif bool(int(database_entry["real"])):
+            dtype = np.dtype(np.float_)
+        else:
+            dtype = np.dtype(np.complex_)
+
+        return cls(
+            matid=database_entry["matid"],
+            group=database_entry["group"],
+            dtype=dtype,
+            nrows=int(database_entry["nrows"]),
+            ncols=int(database_entry["ncols"]),
+            nnz=int(database_entry["nnz"]),
+            is2d3d=bool(int(database_entry["is2d3d"])),
+            isspd=bool(int(database_entry["isspd"])),
+            psym=float(database_entry["psym"]),
+            nsym=float(database_entry["nsym"]),
+            name=database_entry["name"],
+            kind=database_entry["kind"],
+        )
+
+    def _check_matrix_downloaded(self) -> None:
+        if self.matrix is None:
+            raise RuntimeError(
+                "Matrix has not been downloaded yet. Call self.download() first."
+            )
+
+    def _matvec(self, x):
+        self._check_matrix_downloaded()
+        return self.matrix @ x
+
+    def _matmat(self, X):
+        self._check_matrix_downloaded()
+        return self.matrix @ X
+
+    def download(self, verbose: bool = False) -> None:
```

Comment:
As discussed: It would be nice to extract the download of the `SuiteSparseMatrix` in order to make the object immutable and keep it in a 'valid' state at all times, thus alleviating the need for the `self._check_matrix_downloaded()` calls. This could be achieved by implementing something like a `SuiteSparseMatrixBuilder` object which handles downloading and which offers to inspect metadata without the need for a download. Metadata and the matrix itself can then be given to the `SuiteSparseMatrix` linop by the constructor.

Reply:
For now I simply removed the option to not download the matrix, since there is no pressing need for a querying functionality. Once that is added in the future, the `_download` function can simply be moved into a `SuiteSparseMatrixBuilder`.